fix(query): create QueryObserver with initial options

danielr18 opened the pull request:
## Related Issues

Fixes #1416 

## Summary

- Create an observer just once for each `options`.
- Add a hack for `keepPreviousData`.
- Refactor to be similar to `jotai/urql` impl.
  - `write` function no longer returns a promise.

## Check List

- [x] `yarn run prettier` for formatting code and docs

dai-shi commented on src/query/atomWithQuery.ts:
```diff
@@ -99,25 +99,30 @@ export function atomWithQuery<
-  const observerAtom = atom((get) => {
-    const queryClient = getQueryClient(get)
-    const defaultedOptions = queryClient.defaultQueryOptions<
+  let observer:
```
On second look, I think we can probably just have this in `observerAtom`.
```js
      const options =
        typeof createQuery === 'function' ? createQuery(get) : createQuery
```

danielr18 replied:
I think I tried this, but this test failed `query no-loading with keepPreviousData`
```js
const options = typeof createQuery === 'function' ? createQuery(get) : createQuery
```

dai-shi replied:
Oh, no.. Maybe that was the reason.
I'm still not sure if I understand the problem. So, we need to specify `queryKey` when we first create an observer?

danielr18 replied:
For all tests to pass, the observer can only be created once per query.

If createQuery is a function, and it's used in observerAtom, then a new observer would be created if createQuery has a dependency that changes.

I'm new to Jotai, so I don't know how we could have it read the options, without tracking its dependencies, so that only one observer is created.

dai-shi replied:
> once per query

Does it mean if the `queryKey` is the same? regardless of the other parts of the query.

danielr18 replied:
> So, we need to specify `queryKey` when we first create an observer?

Yes

dai-shi replied:
I think the problem is that the current `atomWithQuery` allows changing queryKey dynamically.
We probably shouldn't allow it.
That is going to be a breaking change in its API. If we were to add such a change, we have also other things to change.

Please let us know if you find any solution with dynamic queryKey, then we can merge it meanwhile.

dai-shi reviewed (changes requested):
Our implementation is pretty complicated. (I hope to re-design for better maintainability.)

Having a test is very nice. Please try keeping `observerAtom`.

dai-shi commented on tests/query/atomWithQuery.test.tsx:
```diff
@@ -722,3 +730,47 @@ describe('error handling', () => {
+        response: { count },
+      },
+    ] = useAtom(countAtom)
+    const cachedQueries = useSyncExternalStore(
```
Please don't use uSES here.

I think you can just expect the result?
At L775, `expect(queryClient.getQueryCache().getAll().length).toBe(1)`.

danielr18 replied:
Good catch, done.

danielr18 commented:
> [fa10081](https://github.com/pmndrs/jotai/commit/fa100817e7664d660e230ec46eed591d70462943) tried to use only `queryKey`, which doesn't work..

Would it work if `queryKeyAtom` is wrapped by a selectAtom where the equalityFn always returns true?
I think that would prevent the observerAtom from recomputing.

dai-shi commented:
> always returns true?

but, then it doesn't work if `queryKey` changes?

danielr18 commented:
> > always returns true?
> 
> 
> 
> but, then it doesn't work if `queryKey` changes?

It would ensure the observer isn't recreated when the query key changes. The new query key would be set regardless in the queryAtom, when `observer.setOptions` is called.

dai-shi commented:
Oh, so `observer.setOptions` can override `queryKey`, but it can't be undefined at the first creation?

> wrapped by a selectAtom where the equalityFn always returns true?

Can you try this then? If it works, we'll think about some refactors. (basically, we don't use `jotai/utils` in `jotai/query`.)

danielr18 commented:
> Oh, so `observer.setOptions` can override `queryKey`, but it can't be undefined at the first creation?
> 
> > wrapped by a selectAtom where the equalityFn always returns true?
> 
> Can you try this then? If it works, we'll think about some refactors. (basically, we don't use `jotai/utils` in `jotai/query`.)

I've implemented something similar to `selectAtom`, and tests are passing.

I removed the use of `queryClient.defaultQueryOptions` because `QueryObserver` does it internally.

https://github.com/TanStack/query/blob/63932597db10afc1c7c992acd10e677f489d5fcd/packages/query-core/src/queryObserver.ts#L156

dai-shi commented:
> I've implemented something similar to `selectAtom`, and tests are passing.

Nice. Please note that `refAtom` is a very bad practice and we should fix it in the future. So, it's best not to use it outside selectAtom.

Now, if our requirement is not to re-create observer unless queryClient changes, then probably WeakMap cache is a better option than `refAtom`. Let me try.

> I removed the use of `queryClient.defaultQueryOptions` because `QueryObserver` does it internally.

Oh, this is great. Thanks.

dai-shi commented on src/query/atomWithQuery.ts:
```diff
@@ -90,125 +88,101 @@ export function atomWithQuery<
+        setResult = null
+        if (unsubscribe) {
+          unsubscribe()
+          // FIXME why does this fail?
+          // unsubscribe = null
```
I wasn't be able to solve this. Maybe I'm missing something. Happy if someone tries it.

dai-shi replied:
Ah, maybe it's Strict Effect's double invocation.

dai-shi reviewed (approved):
After working on this, I noticed the problem is not about caching observer across options, but just keeping previous data should make more sense.
I'm not 100% confident and happy with the current implementation, but it's simpler and closer to `jotai/urql` and would be a good path for future refactoring.
There's a little behavioral change, which is more or less a fix in this context.

Thanks for your investigation and work! Without them, this wouldn't have happened.
